### DISCUSSION TITLE
x:Bind vs TemplateBinding inside ControlTemplate

### DISCUSSION THREAD

**Discussion by Naveen61097** [microsoft/microsoft-ui-xaml#9796] [Q&A]
https://learn.microsoft.com/en-us/windows/uwp/xaml-platform/templatebinding-markup-extension#xbind-in-controltemplate

What is the difference between using TemplateBinding and x:Bind inside the ControlTemplate? On the document, it is mentioned that Function Binding and Twoway binding can be done through x:Bind. But is there any internal info that TemplateBinding is more lightweight than x:Bind? Is TemplateBinding better than x:Bind in terms of performance?

**Comment by hoshiizumiya** [comment]
TemplateBinding is one-way mode, reduce the code behind generating and using, which makes it more lightweight.This is only my thought.I still have the same question.

**Comment by ghost1372** [comment]
**TemplateBinding**

**Purpose:** Special lightweight binding used only inside ControlTemplates to forward values from the templated parent (the control).
**Binding Mode:** Always OneWay (cannot be TwoWay).
**Performance:** Extremely lightweight
**Limitations:**

- No converters.
- No functions.
- No TwoWay binding.

**Use Case:** Best when you just need to mirror a control’s property in its template (e.g., Foreground, Background).

**x:Bind**

**Purpose:** Compiled binding — works both inside and outside templates.

**Binding Mode:** Defaults to OneTime, but supports OneWay and TwoWay explicitly.

**Performance:** Faster than standard {Binding} (since it’s compiled), but heavier than TemplateBinding in ControlTemplates.

**Features:**

- Supports functions and expressions.
- Supports converters.
- Can bind to named elements in the template.

**Use Case:** Best when you need logic, converters, or TwoWay binding in a template.

**Limitations:**

- Cannot be used inside a ResourceDictionary and requires a code-behind file.

**Reply by HO-COOH** [reply]
x:Bind **CAN** be used in a `ResourceDictionary` as long as you make a code-behind for it. (Which usually means you create a custom class inheriting from `ResourceDictionary`)

**Comment by AndrewKeepCoding** [comment]
For `TwoWay` binding inside `ControlTemplate`s:

```xaml
<ControlTemplate TargetType="SomeCustomControl">
    <Grid>
        <TextBox Text="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=Text, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
    </Grid>
</ControlTemplate>
```
